poc/observe.py

```diff
@@ -0,0 +1,764 @@
+            equivalent ``filter`` compare equal.
+            i.e. this class should not be exposed to the users.
+        """
+        self.filter = filter
+        self.notify = notify
+
+    def event_factory(self, object, name, old, new):
+        return ObserverEvent(object, name, old, new)
+
+    def __eq__(self, other):
+        if other is self:
+            return True
+        if type(other) is not type(self):
+            return False
+        return (
+            (self.filter, self.notify)
+            == (other.filter, other.notify)
+        )
+
+    def iter_this_targets(self, object):
+        # object must be an instance of HasTraits
+        for name, trait in object.traits().items():
+            if self.filter(name, trait):
+                yield object._trait(name, 2)
+
+    def iter_next_targets(self, object):
+        for name, trait in object.traits().items():
+            if self.filter(name, trait):
+                value = object.__dict__.get(name, Undefined)
+                if is_notifiable(value):
+                    yield value
+
+    def get_change_notifier(self, callback, path, target, dispatcher):
+        return ListenerChangeNotifier(
+            listener_callback=self.change_callback,
+            actual_callback=callback,
+            path=path,
+            target=target,
+            event_factory=self.event_factory,
+            dispatcher=dispatcher,
+        )
+
+    @staticmethod
+    def change_callback(event, callback, path, target, dispatcher):
+        if event.old is not Uninitialized and event.old is not Undefined:
+            remove_notifiers(
+                object=event.old,
+                callback=callback,
+                path=path,
+                target=target,
+                dispatcher=dispatcher,
+            )
+        add_notifiers(
+            object=event.new,
+            callback=callback,
+            path=path,
+            target=target,
+            dispatcher=dispatcher,
+        )
+
+
+class _MetadataFilter:
+    """ Callable as a filter in `FilteredTraitListener` for
+    listening to traits with/without a given metadata.
+    """
+
+    def __init__(self, metadata_name, include):
+        """
+        metadata_name : str
+            Name of metadata.
+        incude : boolean
+            If true, listen to the trait that **have** the metadata.
+            If false, listen to the trait that **do not have** the
+            metadata.
+        """
+        self.metadata_name = metadata_name
+        self.include = include
+
+    def __call__(self, _, trait):
+        return hasattr(trait, self.metadata_name) is self.include
+
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if type(self) is not type(self):
+            return False
+        return (
+            (self.metadata_name, self.include)
+            == (other.metadata_name, other.include)
+        )
+
+
+def MetadataTraitListener(metadata_name, notify, include):
```

Comment:
I think that there might be a better way of implementing this independently of `_FilteredTraitListener` which is more natural for metadata (the `_FilteredTraitListener` class may be independently useful, but I'd make it part of the API).  I'd have the API look something like:
```
from traits.trait_base import not_none

class MetadataTraitListener(BaseListener):
    def __init__(self, metadata_name, notify, filter=not_none):
        ...

   def iter_this_targets(self, object):
        # object must be an instance of HasTraits
        for name in object.trait_name(self.metdata_name=self.filter).items():
            yield object._trait(name, 2)

    ... # etc
```
This is more in line with the way that expert users are used to filtering traits.